### PR TITLE
report: DRY up audit & opportunity rendering

### PR DESCRIPTION
**Pull request by paulirish**
Gets rid of a lot of duplicated work, and brings a lovely consistency to the styles.

https://lighthouse-oppoasanaudit.now.sh/

### CODE REVIEW

File: lighthouse-core/report/html/renderer/util.js [hunk 1]
```diff
@@ -35,9 +35,11 @@ class Util {
    * @return {string}
    */
   static formatDisplayValue(displayValue) {
-    if (typeof displayValue === 'undefined') return '';
     if (typeof displayValue === 'string') return displayValue;
+    if (!displayValue) return '';
 
+    // Don't mutate the value
+    displayValue = [...displayValue];
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/util.js]: oh heh, I guess we might want to use them more than once huh :) thanks for the cleanup!

wanna just remove the mutations? `const template = displayValue[0]; ... of displayValue.slice(1)`

File: lighthouse-core/report/html/renderer/category-renderer.js [hunk 2]
```diff
@@ -26,25 +26,28 @@ class CategoryRenderer {
   /**
    * @param {!ReportRenderer.AuditJSON} audit
    * @param {number} index
+   * @param {DocumentFragment=} providedTmpl
    * @return {!Element}
    */
-  renderAudit(audit, index) {
-    const tmpl = this.dom.cloneTemplate('#tmpl-lh-audit', this.templateContext);
+  renderAudit(audit, index, providedTmpl) {
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/category-renderer.js]: feels like this sorta becomes `populateAuditValues` instead, is that fair?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 3]
```diff
@@ -47,41 +47,27 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
-    this.dom.find('.lh-load-opportunity__wasted-stat', tmpl).textContent = wastedMs;
-    this.dom.find('.lh-load-opportunity__description', tmpl).appendChild(auditDescription);
-
-    // If there's no `type`, then we only used details for `summary`
-    if (details.type) {
-      element.appendChild(this.detailsRenderer.render(details));
+    this.dom.find('.lh-sparkline__bar', element).style.width = sparklineWidthPct;
+    wastedEl.textContent = Util.formatSeconds(summaryInfo.wastedMs, 0.01);
+
+    if (audit.result.displayValue) {
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: so if we're overriding displayValue anyhow and there's no helpText, we're using renderAudit for 

- populating index
- setting rating class
- sticking it inside of details
- filling error/debugString
- something else I'm missing?

WDYT about extracting those 4 to a method they both share?

File: lighthouse-core/report/html/report-styles.css [hunk 4]
```diff
@@ -296,27 +298,25 @@
...
 .lh-audit__header:hover {
   background-color: #F8F9FA;
 }
 
 
 .lh-audit-group[open] > .lh-audit-group__summary > .lh-toggle-arrow,
 .lh-expandable-details[open] > .lh-expandable-details__summary > .lh-toggle-arrow,
-.lh-expandable-details[open] > .lh-expandable-details__summary > div > .lh-toggle-arrow {
+.lh-expandable-details[open] > .lh-expandable-details__summary > div > .lh-toggle-arrow,
+.lh-expandable-details[open] > .lh-expandable-details__summary > div > div > .lh-toggle-arrow {
```
Comment (patrickhulce) [inline lighthouse-core/report/html/report-styles.css]: the nest grows deeper :( removed in chevron ✨ though?

File: lighthouse-core/report/html/renderer/category-renderer.js (hunk 2 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/category-renderer.js]: essentially yeah.

File: lighthouse-core/report/html/report-styles.css (hunk 4 above)
Comment (paulirish) [reply lighthouse-core/report/html/report-styles.css]: yah in #5137 this gets cleaned up a lot.

**Comment by paulirish** [comment]
> not 100% sure what this buys us functionality-wise, slightly nervous we'll get fancier with the opportunities and need to fork again, but I'm probably missing the grand vision

There's been a lot of bugs and inconsistencies recently between opps and audits, all stemming from their separate implementations:

1. opps's toggle-arrow wasn't animating
1. error'd opps werent being handled the same as error'd audits. They needed the "Error!" displaytext + tooltip.
1. Different margin/paddings all around. around the description. on the left/right side of the details tables.
1. debugString placement was different

(To be fair, all these are my fault, but all of these made it past review into master, so... they're sneaky)

In what I have here the templates are defined separately, we're just using the same classes for all the same things. And so we improve maintainability and reduce the duplication.
➡️ ➡️ ➡️  image


net negative what what

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 3 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: i like the sound of that. will do.

**Comment by patrickhulce** [comment]
> There's been a lot of bugs and inconsistencies recently between opps and audits, all stemming from their separate implementations

cool cool, yeah sg with your planned changes 👍

File: lighthouse-core/audits/first-meaningful-paint.js [hunk 5]
```diff
@@ -15,8 +15,8 @@ class FirstMeaningfulPaint extends Audit {
   static get meta() {
     return {
       name: 'first-meaningful-paint',
-      description: 'First meaningful paint',
-      helpText: 'First meaningful paint measures when the primary content of a page is visible. ' +
+      description: 'First Meaningful Paint',
```
Comment (brendankenny) [inline lighthouse-core/audits/first-meaningful-paint.js]: is it possible to squash/rebase/something off of #5130? definitely confusing having all of that in here when it's going to merge with it

File: lighthouse-core/report/html/renderer/category-renderer.js (hunk 2 above)
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/category-renderer.js]: docstring would be good to explain that you can provide an alternative template, just needs to match `#tmpl-lh-audit` in structures a, b, and c

File: lighthouse-core/report/html/renderer/util.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/report/html/renderer/util.js]: add a `util-test.js` test for mutation?

File: lighthouse-core/audits/first-meaningful-paint.js (hunk 5 above)
Comment (paulirish) [reply lighthouse-core/audits/first-meaningful-paint.js]: rebased. much cleaner.

**Comment by paulirish** [comment]
(this is ready for another look)

File: lighthouse-core/report/html/renderer/category-renderer.js [hunk 6]
```diff
@@ -30,21 +30,33 @@ class CategoryRenderer {
...
+  /**
+   * @param {!ReportRenderer.AuditJSON} audit
+   * @param {number} index
+   * @param {!DocumentFragment} tmpl
+   * @return {!Element}
+   */
+  populateAuditValues(audit, index, tmpl) {
     const auditEl = this.dom.find('.lh-audit', tmpl);
     auditEl.id = audit.result.id;
+    const displayTextEl = this.dom.find('.lh-audit__display-text', auditEl);
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/category-renderer.js]: nit: move this into the conditional where it's used?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 7]
```diff
@@ -45,43 +45,29 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
    * @return {!Element}
    */
   _renderOpportunity(audit, index, scale) {
-    const tmpl = this.dom.cloneTemplate('#tmpl-lh-opportunity', this.templateContext);
-    const element = this.dom.find('.lh-load-opportunity', tmpl);
-    element.classList.add(`lh-load-opportunity--${Util.calculateRating(audit.result.score)}`);
+    const oppTmpl = this.dom.cloneTemplate('#tmpl-lh-opportunity', this.templateContext);
+    const element = this.populateAuditValues(audit, index, oppTmpl);
+    element.classList.add(`lh-audit--${Util.calculateRating(audit.result.score)}`);
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: looks like this is now already set in `populateAuditValues` by `_setRatingClass()`

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 8]
```diff
@@ -45,43 +45,29 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
-    this.dom.find('.lh-sparkline__bar', tmpl).style.width = sparklineWidthPct;
-    this.dom.find('.lh-load-opportunity__wasted-stat', tmpl).textContent = wastedMs;
-    this.dom.find('.lh-load-opportunity__description', tmpl).appendChild(auditDescription);
-
-    // If there's no `type`, then we only used details for `summary`
-    if (details.type) {
-      const detailsElem = this.detailsRenderer.render(details);
-      detailsElem.classList.add('lh-details');
-      element.appendChild(detailsElem);
+    const wastedEl = this.dom.find('.lh-audit__display-text', element);
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: maybe comment that everything below here is overwriting the default `displayValue` display provided by `populateAuditValues` with opportunity-specific stuff?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 7 above)
Comment (brendankenny) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: lol `populateAuditValues` also has a stray `auditEl.classList.add(`lh-audit--${audit.result.scoreDisplayMode}`);` line right before calling `_setRatingClass`

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 9]
```diff
@@ -45,43 +45,29 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
+    if (audit.result.scoreDisplayMode !== 'error') {
+      const sparklineWidthPct = `${summaryInfo.wastedMs / scale * 100}%`;
+      this.dom.find('.lh-sparkline__bar', element).style.width = sparklineWidthPct;
+      wastedEl.textContent = Util.formatSeconds(summaryInfo.wastedMs, 0.01);
+    }
+
+    if (audit.result.displayValue) {
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: this seems weird to do if it was an `error` `scoreDisplayMode`. Should be nested inside above `error` conditional?

File: lighthouse-core/report/html/renderer/category-renderer.js [hunk 10]
```diff
@@ -30,21 +30,33 @@ class CategoryRenderer {
    */
   renderAudit(audit, index) {
     const tmpl = this.dom.cloneTemplate('#tmpl-lh-audit', this.templateContext);
+    return this.populateAuditValues(audit, index, tmpl);
+  }
+
+  /**
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/category-renderer.js]: still want a docstring here :P

Comment (paulirish) [reply lighthouse-core/report/html/renderer/category-renderer.js]: done

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 9 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: thanks. didnt need the error conditional after all.

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 7 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: hah. thanks for catching both of these

File: lighthouse-core/test/report/html/renderer/performance-category-renderer-test.js [hunk 11]
```diff
@@ -105,15 +105,33 @@ describe('PerfCategoryRenderer', () => {
...
+        score: 0, scoreDisplayMode: 'numeric',
+        rawValue: 100, explanation: 'Yikes!!', title: 'Bug #2',
       },
     };
 
     const fakeCategory = Object.assign({}, category, {auditRefs: [auditWithDebug]});
     const categoryDOM = renderer.render(fakeCategory, sampleResults.categoryGroups);
 
-    const debugEl = categoryDOM.querySelector('.lh-load-opportunity .lh-debug');
+    const debugEl = categoryDOM.querySelector('.lh-audit--load-opportunity .lh-debug');
```
Comment (brendankenny) [inline lighthouse-core/test/report/html/renderer/performance-category-renderer-test.js]: maybe call this `tooltipEl`? (at least I believe that's what it selects now)

File: lighthouse-core/report/html/renderer/category-renderer.js [hunk 12]
```diff
@@ -71,7 +81,7 @@ class CategoryRenderer {
       const textEl = this.dom.find('.lh-audit__display-text', auditEl);
       textEl.textContent = 'Error!';
       textEl.classList.add('tooltip-boundary');
-      const tooltip = this.dom.createChildOf(textEl, 'div', 'lh-error-tooltip-content tooltip');
+      const tooltip = this.dom.createChildOf(textEl, 'div', 'tooltip lh-debug');
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/category-renderer.js]: does it make sense for the tooltip to share the `lh-debug` class?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 13]
```diff
@@ -76,21 +65,17 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
-    this.dom.find('.lh-load-opportunity__wasted-stat', tmpl).textContent = wastedMs;
-    this.dom.find('.lh-load-opportunity__description', tmpl).appendChild(auditDescription);
-
-    // If there's no `type`, then we only used details for `summary`
-    if (details.type) {
-      const detailsElem = this.detailsRenderer.render(details);
-      detailsElem.classList.add('lh-details');
-      element.appendChild(detailsElem);
+    this.dom.find('.lh-sparkline__bar', element).style.width = sparklineWidthPct;
+    displayEl.textContent = Util.formatSeconds(summaryInfo.wastedMs, 0.01);
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: feels weird to overwrite `Error!` with this in the error case. Will it always have a `summaryInfo` if there was a gatherer or audit error?

File: lighthouse-core/report/html/renderer/category-renderer.js (hunk 12 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/category-renderer.js]: yes

**Comment by paulirish** [comment]
I want to thank the Academy, my wife, Daisy the donkey, and big G.

**Pull request merged**